x/concentrated-liquidity/model/msgs.go

```diff
@@ -0,0 +1,98 @@
+package model
+
+import (
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+
+	cltypes "github.com/osmosis-labs/osmosis/v13/x/concentrated-liquidity/types"
+	swaproutertypes "github.com/osmosis-labs/osmosis/v13/x/swaprouter/types"
+)
+
+// constants.
+const (
+	TypeMsgCreateConcentratedPool = "create_concentrated_pool"
+)
+
+var (
+	_ sdk.Msg                       = &MsgCreateConcentratedPool{}
+	_ swaproutertypes.CreatePoolMsg = &MsgCreateConcentratedPool{}
+)
+
+func NewMsgCreateConcentratedPool(
+	sender sdk.AccAddress,
+	denom0 string,
+	denom1 string,
+	tickSpacing uint64,
+) MsgCreateConcentratedPool {
+	return MsgCreateConcentratedPool{
+		Sender:      sender.String(),
+		Denom0:      denom0,
+		Denom1:      denom1,
+		TickSpacing: tickSpacing,
+	}
+}
+
+func (msg MsgCreateConcentratedPool) Route() string { return cltypes.RouterKey }
+func (msg MsgCreateConcentratedPool) Type() string  { return TypeMsgCreateConcentratedPool }
+func (msg MsgCreateConcentratedPool) ValidateBasic() error {
+	_, err := sdk.AccAddressFromBech32(msg.Sender)
+	if err != nil {
+		return sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "Invalid sender address (%s)", err)
+	}
+
+	if msg.TickSpacing <= 0 {
+		return fmt.Errorf("tick spacing must be positive")
+	}
+
+	if msg.Denom0 == msg.Denom1 {
+		return fmt.Errorf("denom0 and denom1 must be different")
+	}
+
+	if msg.Denom0 == "" || msg.Denom1 == "" {
+		return fmt.Errorf("denom0 and denom1 must be non-empty")
+	}
+
+	return nil
+}
+
+func (msg MsgCreateConcentratedPool) GetSignBytes() []byte {
+	return sdk.MustSortJSON(ModuleCdc.MustMarshalJSON(&msg))
+}
+
+func (msg MsgCreateConcentratedPool) GetSigners() []sdk.AccAddress {
+	sender, err := sdk.AccAddressFromBech32(msg.Sender)
+	if err != nil {
+		panic(err)
+	}
+	return []sdk.AccAddress{sender}
+}
+
+/// Implement the CreatePoolMsg interface
+
+func (msg MsgCreateConcentratedPool) PoolCreator() sdk.AccAddress {
+	sender, err := sdk.AccAddressFromBech32(msg.Sender)
+	if err != nil {
+		panic(err)
+	}
+	return sender
+}
+
+func (msg MsgCreateConcentratedPool) Validate(ctx sdk.Context) error {
+	return msg.ValidateBasic()
+}
```

Comment:
Lets add a TODO for ensuring that the denoms actually exist on-chain

Reply:
https://github.com/osmosis-labs/osmosis/issues/3723